org.eclipse.xtext/src/org/eclipse/xtext/validation/INamesAreUniqueValidationHelper.java

```diff
@@ -8,29 +8,142 @@
  *******************************************************************************/
 package org.eclipse.xtext.validation;
 
+import java.util.Collection;
+
+import org.eclipse.emf.ecore.EClass;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.resource.Resource;
 import org.eclipse.xtext.resource.IEObjectDescription;
+import org.eclipse.xtext.resource.IResourceDescription;
+import org.eclipse.xtext.resource.IResourceDescriptions;
+import org.eclipse.xtext.resource.ISelectable;
+import org.eclipse.xtext.resource.IResourceDescription.Delta;
+import org.eclipse.xtext.resource.impl.IsAffectedExtension;
+import org.eclipse.xtext.scoping.ICaseInsensitivityHelper;
 import org.eclipse.xtext.util.CancelIndicator;
 
 import com.google.inject.ImplementedBy;
 
 /**
+ * The helper can be used to validate unique names locally or across resource boundaries.
+ * 
+ * @see LocalUniqueNameContext
  * @author Sebastian Zarnekow - Initial contribution and API
  */
 @ImplementedBy(NamesAreUniqueValidationHelper.class)
 public interface INamesAreUniqueValidationHelper {
 
+	/**
+	 * Provide the context information for the unique name validation.
+	 * 
+	 * @since 2.22
+	 */
+	@ImplementedBy(DefaultUniqueNameContext.ExportedFromResource.class)
+	interface ContextProvider extends IsAffectedExtension {
+		
+		/**
+		 * Obtain the context of the unique name validation for the given resource.
+		 * 
+		 * May be return null.
+		 */
+		Context tryGetContext(Resource resource, CancelIndicator cancelIndicator);
+		
+		@Override
+		default boolean isAffected(Collection<Delta> deltas, IResourceDescription candidate,
+				IResourceDescriptions context) {
+			return false;
+		}
+	}
+	
+	/**
+	 * Provides context information to the validation that allows to introspect the scope in which
+	 * the validation for unique names is supposed to happen.
+	 * 
+	 * The term <code>clusterType</code> is used to denote a EClass (super-)type of the current object
+	 * under validation. When checking the name for uniqueness, the objects that fulfil the given cluster
+	 * type are considered. This can be used to define different namespaces. For example, a language that
+	 * has the concepts of <code>Fields</code>, <code>Procedures</code> and 
+	 * <code>Functions</code> may provide two different cluster types. All fields must have a unique name
+	 * and the executable <code>Procedures</code> and <code>Functions</code> use the same namespace and may
```

Comment:
is it Procedures _or_ Functions here? (there are so many and's here in the sentence anyway)